src/lib/operators/table_scan.cpp

```diff
@@ -0,0 +1,231 @@
+#include "table_scan.hpp"
+#include "get_table.hpp"
+#include "resolve_type.hpp"
+#include "storage/abstract_attribute_vector.hpp"
+#include "storage/table.hpp"
+#include "storage/value_segment.hpp"
+#include "types.hpp"
+
+namespace opossum {
+
+TableScan::TableScan(const std::shared_ptr<const AbstractOperator>& in, const ColumnID column_id,
+                     const ScanType scan_type, const AllTypeVariant search_value)
+    : AbstractOperator{in}, _column_id{column_id}, _scan_type{scan_type}, _search_value{search_value} {}
+
+ColumnID TableScan::column_id() const {
+  return _column_id;
+}
+
+ScanType TableScan::scan_type() const {
+  return _scan_type;
+}
+
+const AllTypeVariant& TableScan::search_value() const {
+  return _search_value;
+}
+
+std::shared_ptr<const Table> TableScan::_on_execute() {
+  const auto input_table = _left_input_table();
+
+  Assert(input_table, "Performing a table scan without input does not work.");
+  const auto chunk_count = input_table->chunk_count();
+  const auto column_type = input_table->column_type(_column_id);
+  auto output_reference_segments = std::vector<std::shared_ptr<ReferenceSegment>>{};
+
+  // any comparison with null will always return an empty set
+  if (variant_is_null(_search_value)) {
+    return std::make_shared<Table>(*input_table, output_reference_segments);
+  }
+
+  // We might end up using far less segments, but still should be worth to reserve the space for worst case.
+  output_reference_segments.reserve(chunk_count);
+
+  resolve_data_type(column_type, [&](auto type) {
+    using Type = typename decltype(type)::type;
+
```

Comment:
Here could be a nice spot to check whether the search value has the same type as your column as we don't want to compare different types with each other.
```suggestion
Assert(typeid(Type) == _search_value.type(), "Search value doesn't have the same type as the column, we want to compare it with.");
```